chalk-engine/src/logic.rs

```diff
@@ -1006,14 +1028,33 @@ impl<C: Context> Forest<C> {
                             selected_subgoal: _,
                         } = strand;
 
-                        // If the answer had delayed literals, we have to
-                        // ensure that `ex_clause` is also delayed. This is
+                        // If the answer was ambiguous, we have to
+                        // ensure that `ex_clause` is also answer. This is
                         // the SLG FACTOR operation, though NFTD just makes it
                         // part of computing the SLG resolvent.
-                        {
-                            let answer = self.answer(subgoal_table, answer_index);
-                            if answer.ambiguous {
-                                ex_clause.ambiguous = true;
+                        let answer = self.answer(subgoal_table, answer_index);
+                        if answer.ambiguous {
+                            ex_clause.ambiguous = true;
+                        }
+
+                        // For of the delayed subgoals for the answer, we try to check
+                        // if this is a trivial self-cycle. If so, we don't need to add
+                        // this as a delayed_subgoal. Otherwise, we do.
+                        for delayed_subgoal in answer.delayed_subgoals.clone() {
+                            let table = self.stack[depth].table;
+                            let table_goal = &self.tables[table].table_goal;
+                            dbg!(&delayed_subgoal);
+                            dbg!(&table_goal);
+                            if delayed_subgoal == *table_goal {
```

Comment:
Right now, I'm only checking for equality. This is obviously not sufficient (see `cyclic_wf_requirements` test). What's the best way to do this?

Reply:
This has been moved to `goals_trivially_equal` but the same comment applies.